packages/Search/src/details/DTK_DetailsDistributor.hpp

```diff
@@ -28,65 +28,63 @@ namespace DataTransferKit
-  Kokkos::deep_copy(tmp_ranks, destination_ranks);
-  int offset = 0;
-  // this implements a "sort" which is O(N * R) where (R) is
-  // the total number of unique destination ranks.
-  // it performs better than other algorithms in
-  // the case when (R) is small, but results may vary
-  while (true) {
-    int next_biggest_rank;
-    //Kokkos::parallel_reduce("find next biggest rank", n, KOKKOS_LAMBDA(ST i, int& local_max) {
-    //  auto r = tmp_ranks(i);
-    //  local_max = (r > local_max) ? r : local_max;
-    //}, Kokkos::Experimental::Max<int, Kokkos::HostSpace>(next_biggest_rank));
-    next_biggest_rank = *std::max_element(tmp_ranks.data(), tmp_ranks.data() + n);
-    if (next_biggest_rank == -1) break;
-    unique_ranks.push_back(next_biggest_rank);
-    offsets.push_back(offset);
-    Kokkos::View<int, Kokkos::HostSpace> total("total");
-    Kokkos::parallel_scan("process biggest rank items", Kokkos::RangePolicy<Kokkos::Serial>(0, n),
-    KOKKOS_LAMBDA(ST i, int& index, const bool last_pass) {
-      if (last_pass && (tmp_ranks(i) == next_biggest_rank)) {
-        permutation(i) = index + offset;
-      }
-      if (tmp_ranks(i) == next_biggest_rank) ++index;
-      if (last_pass) {
-        if (i + 1 == tmp_ranks.extent(0)) {
-          total() = index;
-        }
-        if (tmp_ranks(i) == next_biggest_rank) {
-          tmp_ranks(i) = -1;
+// Computes the array of indices that sort the input array (in reverse order)
+// but also returns the sorted unique elements in that array with the
+// corresponding element counts and displacement (offsets)
+template <typename InputView, typename OutputView>
+static void sortAndDetermineBufferLayout( InputView ranks,
+                                          OutputView permutation_indices,
+                                          std::vector<int> &unique_ranks,
+                                          std::vector<int> &counts,
+                                          std::vector<int> &offsets )
+{
+    DTK_REQUIRE( unique_ranks.empty() );
+    DTK_REQUIRE( offsets.empty() );
+    DTK_REQUIRE( counts.empty() );
+    DTK_REQUIRE( permutation_indices.extent( 0 ) == ranks.extent( 0 ) );
+    static_assert(
+        std::is_same<typename InputView::non_const_value_type, int>::value,
+        "" );
+    static_assert( std::is_same<typename OutputView::value_type, int>::value,
+                   "" );
+    static_assert(
+        Kokkos::Impl::MemorySpaceAccess<typename OutputView::memory_space,
+                                        Kokkos::HostSpace>::accessible,
+        "" );
+
+    offsets.push_back( 0 );
+
+    auto const n = ranks.extent_int( 0 );
+    if ( n == 0 )
+        return;
+
+    Kokkos::View<int *, Kokkos::HostSpace> ranks_duplicate(
+        Kokkos::ViewAllocateWithoutInitializing( ranks.label() ),
+        ranks.size() );
+    Kokkos::deep_copy( ranks_duplicate, ranks );
+
+    while ( true )
+    {
+        // TODO consider replacing with parallel reduce
```

Comment:
Do you have any plan to change this or should we just remove this code and use `std::sort`